### ISSUE TITLE
Clarify the purpose of the fail-fast option

### ISSUE THREAD

**Issue by drnextgis** [stac-utils/stac-asset#201]
I'm using the following command with the `--fail-fast` option. According to the documentation, I expect it to fail immediately on encountering a download error, rather than waiting for all downloads to complete. However, I'm still getting multiple errors like this, and the command continues running. Am I misunderstanding the purpose of this argument?
```
$ stac-asset -v DEBUG download --fail-fast --path-template '${satl:outcome_id}' --http-timeout 1
```

```
ERROR: SL-newsat26-20240714T082254933086Z-micro-L0-v0.5.8[analytic] - TimeoutError:                                                                                                                                
131/140:   2%|██▏                                                                                                                                         | 51.5M/3.28G [00:00<00:02, 1.29GB/s, 27 errors, 0 skips]Task exception was never retrieved
future: <Task finished name='Task-248' coro=<Downloads._download_with_release() done, defined at /tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py:182> exception=TimeoutError()>
Traceback (most recent call last):
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 186, in _download_with_release
    return await download.download(messages=messages)
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 63, in download
    raise error
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 53, in download
    await download_asset(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 423, in download_asset
    raise error
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 411, in download_asset
    await client.download_href(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/client.py", line 120, in download_href
    raise err
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/client.py", line 102, in download_href
    async for chunk in self.open_href(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/client.py", line 78, in open_href
    async for chunk in self.open_url(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/http_client.py", line 165, in open_url
    async for chunk, _ in response.content.iter_chunks():
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/streams.py", line 66, in __anext__
    rv = await self._stream.readchunk()
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/streams.py", line 441, in readchunk
    await self._wait("readchunk")
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/streams.py", line 311, in _wait
    with self._timer:
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/helpers.py", line 735, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
ERROR: SL-newsat26-20240714T082254918942Z-micro-L0-v0.5.8[analytic] - TimeoutError:                                                                                                                                
133/142:   2%|██▏                                                                                                                                          | 53.3M/3.37G [00:00<00:06, 540MB/s, 28 errors, 0 skips]Task exception was never retrieved
future: <Task finished name='Task-256' coro=<Downloads._download_with_release() done, defined at /tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py:182> exception=TimeoutError()>
Traceback (most recent call last):
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 186, in _download_with_release
    return await download.download(messages=messages)
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 63, in download
    raise error
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 53, in download
    await download_asset(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 423, in download_asset
    raise error
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/_functions.py", line 411, in download_asset
    await client.download_href(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/client.py", line 120, in download_href
    raise err
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/client.py", line 102, in download_href
    async for chunk in self.open_href(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/client.py", line 78, in open_href
    async for chunk in self.open_url(
  File "/tmp/env/lib/python3.10/site-packages/stac_asset/http_client.py", line 165, in open_url
    async for chunk, _ in response.content.iter_chunks():
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/streams.py", line 66, in __anext__
    rv = await self._stream.readchunk()
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/streams.py", line 441, in readchunk
    await self._wait("readchunk")
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/streams.py", line 311, in _wait
    with self._timer:
  File "/tmp/env/lib/python3.10/site-packages/aiohttp/helpers.py", line 735, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
```

Or does `--fail-fast` only apply to aiohttp errors when using the HTTP client?

**Comment by drnextgis** [comment]
Is this:
```
Task exception was never retrieved
```
something we should care about?

**Comment by gadomski** [comment]
> Or does --fail-fast only apply to aiohttp errors when using the HTTP client?

No, it _should_ work for any set of asset downloads. If we're getting multiple errors coming back to the CLI, that's a bug. These things are tricky to debug without a real-world example, so if you're able to dig in on your side that might help a lot. The critical logic will be in https://github.com/stac-utils/stac-asset/blob/e29561105372765160c5dacf133ba5a663863812/src/stac_asset/_functions.py#L146-L181 ... I'd be curious what was happening in that block from 155-163, e.g., to make sure we're really just throwing a single error, etc...

**Comment by drnextgis** [comment]
@gadomski here are the steps to reproduce the issue with a publicly available STAC catalog:

1. Set the `max_concurrent_downloads` [here](https://github.com/stac-utils/stac-asset/blob/main/src/stac_asset/_functions.py#L84) to `5`. This setting is important for reproducing the issue. I was able to reproduce the issue with `max_concurrent_downloads` less than number of assets for the given STAC item. 
2. Attempt to download a STAC item with a reduced `--http-timeout` set to `5` to trigger errors:
   ```shell
    stac-asset download --http-timeout 5 --fail-fast \
    https://landsatlook.usgs.gov/stac-server/collections/landsat-c2l2alb-ta/items/LC09_L2SP_070011_20240717_20240719_02_A1_TOA
   ```

**Comment by drnextgis** [comment]
@gadomski, were you able to reproduce the issue on your side?

**Comment by gadomski** [comment]
Yup, I _think_ I have a fix in #204, if you don't mind taking a look.

**Comment by drnextgis** [comment]
Great! I'll check it out. Thanks for adding the `--max-concurrent-downloads` option to the CLI 👍
